### DISCUSSION TITLE
Proper way to cancel effects?

### DISCUSSION THREAD

**Discussion by sillygoose** [pointfreeco/swift-composable-architecture#2450] [Q&A] [answered]
Wondering the proper way to cancel effects when a feature is dismissed.  The child feature will run various network queries or timers and when the feature is dismissed I would cancel any in-flight queries or timers.

The question is where in the child feature can I put the cancel code?  If I use `.onDisappear()` the view is gone and I will get warnings like:
```
An "ifLet" at "MyRivian/Trips.swift:496" received a presentation action when destination state was absent. …

  Action:
    Trips.Action.destination(.presented(.monthlyEfficiency(.onDisappear)))

This is generally considered an application logic error, and can happen for a few reasons:
```
There isn't a SwiftUI-equivalent to the UIKit `viewWillDisappear` event so wondering how I can see that the feature is being dismissed and cancel any in-flight effects before it has disappeared.  I did find some suggestions to drop down to UIKit and create a `onWillDisappear()` event but wondering if there are other ways.

**Comment by mbrandonw** [comment, marked as answer]
Hi @sillygoose, are you using the presentation tools that come with the library, such as `PresentationState`, `PresentationAction`, `ifLet`, etc? If so, those tools automatically cancel effects when a feature is dismissed. And if not, is there a reason not to?

**Reply by sillygoose** [reply]
I am changing everything over to the new presentation tools so that will be a nice side effect, one less issue to deal with.

**Comment by acosmicflamingo** [comment]
I 1000% love your username @sillygoose . Would you like to swap?

**Reply by sillygoose** [reply]
I would consider it but my Point-Free account is tied to it and I can't risk missing any episodes right now.

**Reply by acosmicflamingo** [reply]
Your response absolutely lives up to your username. Thank you and godspeed my good ~~man~~ goose.
